### PR TITLE
fix+refactor: asyncfl loss divergence

### PR DESCRIPTION
**Pull request by myungjin**
## Description

Please provide a meaningful description of what this change will do, or is for. Bonus points for including links to
related issues, other PRs, or technical references.

For asyncfl, a client (trainer) should send delta by subtracting local weights from original global weights after training. In the current implementation, the whole local weights were sent to a server (aggregator). This causes loss divergence.

Supporting delta update requires refactoring of aggregators of synchronous fl (horizontal/{top_aggregator.py, middle_aggregator.py}) as well as optimizers' do() function.

The changes here support delta update universally across all types of modes (horizontal synchronous, asynchronous, and hybrid).

## Type of Change

- [x] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [x] Refactor
- [ ] Documentation
- [ ] Other (please describe)

### CODE REVIEW

File: lib/python/flame/common/typing.py [hunk 1]
```diff
@@ -13,11 +13,12 @@
...
-
 """Definitions on Types."""
 
 from typing import Union
 
 Scalar = Union[bool, bytes, float, int, str]
 
 Metrics = dict[str, Scalar]
+
+ModelWeights = Union[list, dict]
```
Comment (elqurio) [inline lib/python/flame/common/typing.py:24]: Not sure what to think about this. It seems wrong.
When I see ModelWeights, I assume it's a list, because of plurality. But it can be a list or a dict.
Maybe we need a ModelWeight type, and then if we want a list of these, we can do list[ModelWeight]

File: lib/python/flame/common/util.py [hunk 2]
```diff
@@ -115,3 +116,21 @@ def mlflow_runname(config: Config) -> str:
...
+    return {x: a[x] - b[y] for (x, y) in zip(a, b)}
+
+
+def delta_weights_tensorflow(a: ModelWeights,
+                             b: ModelWeights) -> Union[ModelWeights, None]:
+    """Return delta weights for tensorflow model weights."""
+    if a is None or b is None:
+        return None
+
+    return [x - y for (x, y) in zip(a, b)]
```
Comment (elqurio) [inline lib/python/flame/common/util.py:136]: Ah, so for pytorch, the weights are dict, but for tensorflow they are a list.
Seems to me like we can implement a factory here. But the typing now makes more sense.

File: lib/python/flame/examples/hier_mnist/middle_aggregator/config_uk.json [hunk 3]
```diff
@@ -1,10 +1,14 @@
 {
     "taskid": "49d06b7526964db86cf37c70e8e0cdb6bd7aa743",
-    "backend": "mqtt",
+    "backend": "p2p",
```
Comment (elqurio) [inline lib/python/flame/examples/hier_mnist/middle_aggregator/config_uk.json:3]: Are these changes here by mistake? Don't see the connection with this PR

File: lib/python/flame/mode/distributed/trainer.py [hunk 4]
```diff
@@ -53,16 +55,14 @@ def init_cm(self) -> None:
...
         self.registry_client = registry_provider.get(self.config.registry.sort)
         # initialize registry client
         self.registry_client(self.config.registry.uri, self.config.job.job_id)
 
         base_model = self.config.base_model
         if base_model and base_model.name != "" and base_model.version > 0:
             self.model = self.registry_client.load_model(
-                base_model.name, base_model.version
-            )
+                base_model.name, base_model.version)
```
Comment (elqurio) [inline lib/python/flame/mode/distributed/trainer.py:65]: We should consider adding a auto formatter, like black to prevent these dust changes

File: lib/python/flame/mode/horizontal/asyncfl/middle_aggregator.py [hunk 5]
```diff
@@ -161,6 +176,9 @@ def _aggregate_weights(self, tag: str) -> None:
             time.sleep(1)
             return
 
+        # save global weights before updating it
+        self.prev_weights = self.weights
```
Comment (elqurio) [inline lib/python/flame/mode/horizontal/asyncfl/middle_aggregator.py:180]: Should this be a deepcopy as well?

File: lib/python/flame/mode/horizontal/trainer.py [hunk 6]
```diff
@@ -143,6 +153,9 @@ def _update_model(self):
             self.model.set_weights(self.weights)
 
     def _update_weights(self):
+        # save weights before updating it
+        self.prev_weights = self.weights
```
Comment (elqurio) [inline lib/python/flame/mode/horizontal/trainer.py:157]: Deepcopy?

**Review by elqurio — COMMENTED** [review]
Apart from config changes in examples, the branching for PyTorch and TF, and a couple of deepcopies, it looks good to me.
As for examples, shouldn't we add a different example for asyncFL rather than updating existing ones which are synchronous?

File: lib/python/flame/common/typing.py (hunk 1 above)
Comment (myungjin) [reply lib/python/flame/common/typing.py:24]: This can be confusing. A ML model can be represented as list or dict depending on the framework (pytorch or tensorflow). Each is ModelWeights. The reason I use the plural form rather than a singular form is that weight usually means one value in the list or dict.

File: lib/python/flame/common/util.py (hunk 2 above)
Comment (myungjin) [reply lib/python/flame/common/util.py:136]: a factory pattern can be implemented with design changes, which require more thinking.

File: lib/python/flame/examples/hier_mnist/middle_aggregator/config_uk.json (hunk 3 above)
Comment (myungjin) [reply lib/python/flame/examples/hier_mnist/middle_aggregator/config_uk.json:3]: you're right. They aren't.
Now that we have local setting doc (https://github.com/cisco-open/flame/blob/main/docs/08-flame-sdk.md#quickstart), I wanted to align the examples with the doc.

Instead of creating a separate PR, I piggybacked these into this PR. If you want to separate a PR, I can remove them from here. let me know.

File: lib/python/flame/mode/distributed/trainer.py (hunk 4 above)
Comment (myungjin) [reply lib/python/flame/mode/distributed/trainer.py:65]: I am using python lsp. sometimes, it goes back and forth with different format.
If we can enforce auto-formatting before PR submission, that would be nice. But I don't know how.

File: lib/python/flame/mode/horizontal/asyncfl/middle_aggregator.py (hunk 5 above)
Comment (myungjin) [reply lib/python/flame/mode/horizontal/asyncfl/middle_aggregator.py:180]: No, it's not necessary because I already did deepcopy at https://github.com/cisco-open/flame/pull/330/files#diff-074d31d6330e5c3880f78eef8d7bfbcfd61746fe1de6a990d07cbc48c61ebcafR134.

To summarize the structure,

self._agg_goal_weight = deepcopy(self.weights) // let's say the address of self._agg_goal_weights is `abc` and the address of self.weights is `def`
// do some aggregation with self._agg_goal_weight.

// once some condition is met, we reach the following statement
self.prev_weights = self.weights // addr(prev_weights) = `def`
self.weights = self._agg_goal_weights // addr(weights) = `abc`

File: lib/python/flame/mode/horizontal/trainer.py (hunk 6 above)
Comment (myungjin) [reply lib/python/flame/mode/horizontal/trainer.py:157]: we are safe here too because of https://github.com/cisco-open/flame/pull/330/files#diff-148a2e70215f277621faa325f4be3ab1be7ca28f9729e3732e7ea86c89eca326R97.

self.weights = self.weights = msg[MessageType.WEIGHTS] // let's say addr(weights) = `abc`.
then we call `self._update_model()`, which load weights into a model (e.g., by doing `self.model.load_state_dict(self.weights)` in `self._update_model()`). This won't let weights in model point to `abc`.

So, when we call `self.prev_weights = self.weights` , addr(self.prev_weights) = `abc`. when we call `self.weights = self.model.state_dict()` for example, addr(self.weights) may be some other value than `abc`.

Also, prev_weights and weights point to the same object, delta_weights would have been zero.

File: lib/python/flame/common/util.py (hunk 2 above)
Comment (elqurio) [reply lib/python/flame/common/util.py:136]: Agreed, it's a much bigger change than this PR.

File: lib/python/flame/mode/distributed/trainer.py (hunk 4 above)
Comment (elqurio) [reply lib/python/flame/mode/distributed/trainer.py:65]: I can set up pre commit hooks to run black I think. Also one of the tools I like is isort, it sorts imports in a specific order (stdlib > installed libs > project imports). And it has a profile to work well with black, because black likes to do some formatting on sorts as well.
